atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java

```diff
@@ -0,0 +1,135 @@
+import com.palantir.atlasdb.transaction.impl.TransactionRetryStrategy.Retryable;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TransactionRetryStrategyTest {
+    private final TestBlockStrategy blockStrategy = new TestBlockStrategy();
+    @Mock private ThreadLocalRandom random;
+    @Mock private Retryable<String, Exception> task;
+    @Mock private IntPredicate shouldStopRetrying;
+
+    private TransactionRetryStrategy legacy;
+    private TransactionRetryStrategy exponential;
+
+    private static final class TestBlockStrategy implements BlockStrategy {
+        private long totalBlockedTime;
+
+        @Override
+        public void block(long sleepTime) {
+            totalBlockedTime += sleepTime;
+        }
+    }
+
+    @Before
+    public void before() {
+        when(random.nextLong(anyLong())).thenAnswer(inv -> (long) inv.getArgument(0) - 1);
+        when(shouldStopRetrying.test(anyInt())).thenAnswer(inv -> ((int) inv.getArgument(0)) > 1);
+        legacy = TransactionRetryStrategy.createLegacy(blockStrategy);
+        exponential = TransactionRetryStrategy.createExponential(blockStrategy, random);
+
+    }
+
+    private String run() throws Exception {
+        return exponential.runWithRetry(shouldStopRetrying, task);
+    }
+
+    @Test
+    public void retriesIfFailsWithRetriableException() throws Exception {
+        when(task.run()).thenThrow(new TransactionFailedRetriableException("")).thenReturn("success");
+        assertThat(run()).isEqualTo("success");
+    }
+
+    @Test
+    public void stopsIfShouldStopRetrying() throws Exception {
+        TransactionFailedRetriableException second = new TransactionFailedRetriableException("second");
+        when(task.run()).thenThrow(new TransactionFailedRetriableException("first")).thenThrow(second);
+        assertThatExceptionOfType(TransactionFailedRetriableException.class)
+                .isThrownBy(this::run)
+                .withMessage("Failing after 2 tries.")
+                .withCause(second);
+    }
+
+    @Test
+    public void doesNotRetryOnNonRetriableTransactionFailedException() throws Exception {
+        TransactionFailedNonRetriableException failure = new TransactionFailedNonRetriableException("");
+        when(task.run()).thenThrow(failure);
+        assertThatThrownBy(this::run).isEqualTo(failure);
+    }
+
+    @Test
+    public void rethrowsExceptions() throws Exception {
+        Exception exception = new Exception("rethrown");
+        when(task.run()).thenThrow(exception);
+        assertThatThrownBy(this::run).isEqualTo(exception);
+    }
+
+    @Test
+    public void rethrowsErrors() throws Exception {
+        AssertionError error = new AssertionError();
+        when(task.run()).thenThrow(error);
+        assertThatThrownBy(this::run).isEqualTo(error);
+    }
+
+    @Test
+    public void doesNotBackOffIfLegacy() throws Exception {
+        when(shouldStopRetrying.test(anyInt())).thenAnswer(inv -> ((int) inv.getArgument(0)) > 50);
```

Comment:
not really sure what's going on here, can we move it to a method, so it clears up the individual test, we have it in a few other places in the test